FWCore/Framework/src/PathsAndConsumesOfModules.cc

```diff
@@ -288,196 +349,352 @@ namespace edm {
       if (description->id() > largestIndex) {
         largestIndex = description->id();
       }
+      if (description->moduleName() == kPathStatusInserter) {
+        triggerResultsStatus.dependsOn_.push_back(description->id());
+      }
       if (description->moduleName() == kPathStatusInserter || description->moduleName() == kEndPathStatusInserter) {
         pathStatusInserterModuleLabelToModuleID[description->moduleLabel()] = description->id();
       }
     }
-    kPathToTriggerResultsDependencyLastIndex = largestIndex;
 
-    /*
-    {
-      //We need to explicitly check that modules on Paths do not try to read data from
-      // Modules which are only on EndPaths. The circular dependency finder has been
-      // known to miss these.
-      std::unordered_set<unsigned int> modulesOnlyOnEndPaths;
-      auto const& endPaths = iPnC.endPaths();
-      for( unsigned int pathIndex = 0; pathIndex != endPaths.size(); ++pathIndex) {
-        auto const& moduleDescriptions = iPnC.modulesOnEndPath(pathIndex);
-        for(auto const& description: moduleDescriptions) {
-          modulesOnlyOnEndPaths.insert(description->id());
-        }
-      }
+    std::vector<ModuleStatus> statusOfModules(largestIndex + 1);
+    for (auto const& nameID : pathStatusInserterModuleLabelToModuleID) {
+      statusOfModules[nameID.second].onPath_ = true;
+    }
+    if (kTriggerResultsIndex != kInvalidIndex) {
+      statusOfModules[kTriggerResultsIndex] = std::move(triggerResultsStatus);
+    }
+
+    std::vector<PathStatus> statusOfPaths(iPnC.paths().size() + iPnC.endPaths().size());
 
-      std::unordered_set<unsigned int> modulesOnPaths;
-      auto const& paths = iPnC.paths();
-      for( unsigned int pathIndex = 0; pathIndex != paths.size(); ++pathIndex) {
-        auto const& moduleDescriptions = iPnC.modulesOnPath(pathIndex);
-        for(auto const& description: moduleDescriptions) {
-          auto itFind =modulesOnlyOnEndPaths.find(description->id());
-          if(modulesOnlyOnEndPaths.end() != itFind) {
-            modulesOnlyOnEndPaths.erase(itFind);
+    //If there are no paths, no modules will run so nothing to check
+    if (statusOfPaths.empty()) {
+      return;
+    }
+
+    {
+      auto nPaths = iPnC.paths().size();
+      for (unsigned int p = 0; p < nPaths; ++p) {
+        auto& status = statusOfPaths[p];
+        status.index_ = p;
+        status.modulesOnPath_.reserve(status.nModules_);
```

Comment:
It looks like nModules_ is zero at this point so this reserve does nothing. Same below in the endPaths loop.

Reply:
Good catch. That is a casualty of a later change I made.